test/integration/retryable-writes/retryable_writes.spec.test.ts

```diff
@@ -0,0 +1,209 @@
+        // Step 2: Run the test
+        const mochaTest = it(spec.description, async () => await executeScenarioTest(spec, ctx));
+
+        // A pattern we don't need to repeat for unified tests
+        // In order to give the beforeEach hook access to the
+        // spec test so it can be responsible for skipping it
+        // and executeScenarioSetup
+        mochaTest.spec = spec;
+      }
+    });
+  }
+});
+
+async function executeScenarioSetup(scenario, test, config, ctx) {
+  const url = config.url();
+  const options = {
+    ...test.clientOptions,
+    heartbeatFrequencyMS: 100,
+    monitorCommands: true,
+    minPoolSize: 10
+  };
+
+  ctx.failPointName = test.failPoint && test.failPoint.configureFailPoint;
+
+  const client = config.newClient(url, options);
+  await client.connect();
+
+  ctx.client = client;
+  ctx.db = client.db(config.db);
+  ctx.collection = ctx.db.collection(`retryable_writes_test_${config.name}_${test.operation.name}`);
+
+  try {
+    await ctx.collection.drop();
+  } catch (error) {
+    if (!error.message.match(/ns not found/)) {
+      throw error;
+    }
+  }
+
+  if (Array.isArray(scenario.data) && scenario.data.length) {
+    await ctx.collection.insertMany(scenario.data);
+  }
+
+  if (test.failPoint) {
+    await ctx.client.db('admin').command(test.failPoint);
+  }
+}
+
+async function executeScenarioTest(test, ctx) {
+  const args = generateArguments(test);
+
+  // In case the spec files or our API changes
+  expect(ctx.collection).to.have.property(test.operation.name).that.is.a('function');
+
+  let thrownError;
+  let result;
+  try {
+    result = await ctx.collection[test.operation.name](...args);
+  } catch (error) {
+    thrownError = error;
+  }
+
+  const outcome = test.outcome && test.outcome.result;
+  const errorLabelsContain = outcome && outcome.errorLabelsContain;
+  const errorLabelsOmit = outcome && outcome.errorLabelsOmit;
+  const hasResult = outcome && !errorLabelsContain && !errorLabelsOmit;
+  if (test.outcome.error) {
+    expect(thrownError, `${test.operation.name} was supposed to fail but did not!`).to.exist;
+    expect(thrownError).to.have.property('message');
+
+    if (hasResult) {
+      expect(thrownError.result).to.matchMongoSpec(test.outcome.result);
+    }
+
+    if (errorLabelsContain) {
+      expect(thrownError.errorLabels).to.include.members(errorLabelsContain);
+    }
+
+    if (errorLabelsOmit) {
+      for (const label of errorLabelsOmit) {
+        expect(thrownError.errorLabels).to.not.contain(label);
+      }
+    }
+  } else if (test.outcome.result) {
+    expect(thrownError, thrownError?.stack).to.not.exist;
+    const expected = test.outcome.result;
+    result = transformToResultValue(result);
+    expect(result).to.deep.include(expected);
+  }
+
+  if (test.outcome.collection) {
+    const collectionResults = await ctx.collection.find({}).toArray();
+
+    expect(collectionResults).to.deep.equal(test.outcome.collection.data);
+  }
+
+  return result;
```

Comment:
I also noticed that we aren't asserting on command construction here [like the spec says](https://github.com/mongodb/specifications/tree/master/source/retryable-writes/tests#command-construction-tests) even though we do turn on command monitoring. It may be out of scope for this ticket, but can you check my sanity here and maybe we can file a follow up?

Reply:
Yea I don't see any command collection happening, I think I can do it without much effort but I'll post a ticket if it's not simple.

Reply:
still planning to do this?

Reply:
Filed a ticket: NODE-4033, the complexity of bulkWrites and findAndModify operations should be done in a follow up